apigee_sync.go

```diff
@@ -213,67 +210,95 @@ func pollChangeAgent() error {
 	}
 }
 
+
+// simple doubling back-off
+func createBackOff(retryIn, maxBackOff time.Duration) func() {
+	return func() {
+		log.Debugf("backoff called. will retry in %s.", retryIn)
+		time.Sleep(retryIn)
+		retryIn = retryIn * time.Duration(2)
+		if retryIn > maxBackOff {
+			retryIn = maxBackOff
+		}
+	}
+}
+
 /*
  * This function will (for now) use the Access Key/Secret Key/ApidConfig Id
  * to get the bearer token, and the scopes (as comma separated scope)
  */
-func getBearerToken() bool {
+func getBearerToken() {
 
 	log.Info("Getting a Bearer token...")
-	uri, err := url.Parse(config.GetString(configProxyServerBaseURI))
+	uriString := config.GetString(configProxyServerBaseURI)
+	uri, err := url.Parse(uriString)
 	if err != nil {
-		log.Error(err)
-		return false
+		log.Panicf("unable to parse uri config '%s' value: '%s': %v", configProxyServerBaseURI, uriString, err)
 	}
 	uri.Path = path.Join(uri.Path, "/accesstoken")
 
-	token = ""
-	form := url.Values{}
-	form.Set("grant_type", "client_credentials")
-	form.Add("client_id", config.GetString(configConsumerKey))
-	form.Add("client_secret", config.GetString(configConsumerSecret))
-	req, err := http.NewRequest("POST", uri.String(), bytes.NewBufferString(form.Encode()))
-	req.Header.Set("Content-Type", "application/x-www-form-urlencoded; param=value")
-	req.Header.Set("display_name", apidInfo.InstanceName)
-	req.Header.Set("apid_instance_id", apidInfo.InstanceID)
-	req.Header.Set("apid_cluster_Id", apidInfo.ClusterID)
-	req.Header.Set("status", "ONLINE")
-	req.Header.Set("created_at_apid", time.Now().Format(time.RFC3339))
-	req.Header.Set("plugin_details", apidPluginDetails)
+	retryIn := 5 * time.Millisecond
+	maxBackOff := 1 * time.Minute
+	backOffFunc := createBackOff(retryIn, maxBackOff)
+	first := true
 
-	client := &http.Client{}
-	resp, err := client.Do(req)
-	if err != nil {
-		log.Error("Unable to Connect to Edge Proxy Server ", err)
-		return false
-	}
-	defer resp.Body.Close()
-	if resp.StatusCode != 200 {
-		log.Error("Oauth Request Failed with Resp Code ", resp.StatusCode)
-		return false
-	}
-	body, err := ioutil.ReadAll(resp.Body)
-	if err != nil {
-		log.Error("Unable to read EdgeProxy Sever response ", err)
-		return false
-	}
+	for {
+		if first {
+			first = false
+		} else {
+			backOffFunc()
+		}
 
-	var oauthResp oauthTokenResp
-	log.Debugf("Response: %s ", body)
-	err = json.Unmarshal(body, &oauthResp)
-	if err != nil {
-		log.Error(err)
-		return false
-	}
-	token = oauthResp.AccessToken
+		token = ""
+		form := url.Values{}
+		form.Set("grant_type", "client_credentials")
+		form.Add("client_id", config.GetString(configConsumerKey))
```

Comment:
No. I thought it was just redundant each time. May be looking only at the diff's, it unclear.